book/src/proptest-derive/modifiers.md

```diff
@@ -0,0 +1,370 @@
+See the documentation for [`prop_filter`] for more details.
+
+The argument to the modifier must be a valid argument for the second parameter
+of [`prop_filter`].
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+#[proptest(filter = "|segment| segment.start != segment.end")]
+struct NonEmptySegment {
+    start: i32,
+    end: i32,
+}
+```
+
+As mentioned above, filtering should be avoided when it is reasonably possible
+to express a non-filtering strategy that achieves the same effect. For example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct BadExapmle {
+    // Don't do this! Your tests will run more slowly and shrinking won't work
+    // properly.
+    #[proptest(filter = "|x| x % 2 == 0")]
+    even_number: u32,
+}
+
+#[derive(Debug, Arbitrary)]
+struct GoodExample {
+    // Directly generate even numbers only by transforming the set of all
+    // `u32`s and then mapping it to the set of even `u32`s.
+    #[proptest(strategy = "any::<u32>().prop_map(|x| x / 2 * 2)")]
+    even_number: u32,
+}
+```
+
+[`prop_filter`]: https://altsysrq.github.io/rustdoc/proptest/latest/proptest/strategy/trait.Strategy.html#method.prop_filter
+
+## `no_bound`
+
+Form: `#[proptest(no_bound)]`
+
+Usable on: generic type declarations
+
+Normally, when `#[derive(Arbitrary)]` is applied to an item with generic type
+arguments, every type argument is required to `impl Arbitrary`. For example,
+given a declaration like the following:
+
+```rust
+#[derive(Arbitrary)]
+struct MyStruct<T> { /* ... */ }
+```
+
+Something like this will be generated:
+
+```rust
+impl<T> Arbitrary for MyStruct<T> where T: Arbitrary { /* ... */ }
+```
+
+Placing `#[proptest(no_bound)]` on a generic type argument suppresses this. For
+example, the following removes the extra `where T: Arbitrary`:
+
+```rust
+#[derive(Arbitrary)]
+struct MyStruct<#[proptest(no_bound)] T> { /* ... */ }
+```
+
+In order for `no_bound` to be useful, one of two things needs to be true:
+
+- The nature of the item is such that no value with the generic type actually
+  comes into existence. This can happen if the only references are things such
+  as `PhantomData<T>` or if the item is an enum and all variants referencing
+  the type are marked `#[proptest(skip)]`.
+
+- Other modifiers (such as `value`) are added to things referencing the generic
+  type that remove proptest's responsibility for generating them.
+
+Example of the above:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum MyEnum<#[proptest(no_bound)] T> {
+    // OK -- no T needs to be constructed
+    Phantom(PhantomData<T>),
+
+    // OK -- we're manually taking responsibility for generating the value
+    #[proptest(value = "MyEnum::WithValue(None)")]
+    WithValue(Option<T>),
+
+    // OK -- we tell proptest not to generate this case
+    #[proptest(skip)]
+    Skipped(T),
+
+    // ERROR -- proptest would need to generate a T here, but cannot because
+    // we lack the `T: Arbitrary` bound due to having used `no_bound`.
+    Error(T),
```

Comment:
But this is correct; In general, when you make the macro use `any` or `any_with` then `T: Arbitrary` will be added... You can see this by looking for `.mark_uses(` and some of the logic inside `UseTracker::add_bounds`.

This also applies when you use some associated type... If we we write (`tests/assoc.rs`):
```rust
#[derive(Debug, Arbitrary)]
struct T4<Tyvar: FuncB> {
    field: Tyvar::OutB,
}
```
then we require `<Tyvar as FuncB>::OutB: Arbitrary` but not `Tyvar: Arbitrary`.

The following situation would require `no_bound` tho:

```rust
#[derive(Debug, Arbitrary)]
struct Hide<T>(PhantomData<T>);

#[derive(Debug, Arbitrary)]
struct Stuff<T>(Hide<T>);
// We have no idea that `Arbitrary for Hide<T>` doesn't require `T: Arbitrary`
// so we use `no_bound` to get rid of the extra bound.
```

I should probably add a new round of tests to ensure the semantics are as expected and more robust in the face of changes. Some documentation improvements in the code itself are also probably necessary on my part.